File: lib/src/main/java/pl/tajchert/sample/JumpingSpan.java
package pl.tajchert.sample;

import android.graphics.Canvas;
import android.graphics.Paint;
import android.graphics.Paint.FontMetricsInt;
import android.text.style.ReplacementSpan;

public class JumpingSpan extends ReplacementSpan {

    private float translationX = 0;
    private float translationY = 0;

    @Override
    public int getSize(Paint paint, CharSequence text, int start, int end, FontMetricsInt fontMetricsInt) {
        return (int) paint.measureText(text, start, end);
    }

    @Override
    public void draw(Canvas canvas, CharSequence text, int start, int end, float x, int top, int y, int bottom, Paint paint) {
        canvas.drawText(text, start, end, x + translationX, y + translationY, paint);
    }

    public void setTranslationX(float translationX) {
        this.translationX = translationX;
    }

    public void setTranslationY(float translationY) {
        this.translationY = translationY;
    }
}

File: lib/src/main/java/pl/tajchert/waitingdots/JumpingSpan.kt
package pl.tajchert.waitingdots

import android.graphics.Canvas
import android.graphics.Paint
import android.graphics.Paint.FontMetricsInt
import android.text.style.ReplacementSpan

class JumpingSpan : ReplacementSpan() {

  private var translationX = 0f
  private var translationY = 0f

  override fun getSize(
    paint: Paint,
    text: CharSequence,
    start: Int,
    end: Int,
    fontMetricsInt: FontMetricsInt?
  ): Int {
    return paint.measureText(text, start, end)
        .toInt()
  }

  override fun draw(
    canvas: Canvas,
    text: CharSequence,
    start: Int,
    end: Int,
    x: Float,
    top: Int,
    y: Int,
    bottom: Int,
    paint: Paint
  ) {
    canvas.drawText(text, start, end, x + translationX, y + translationY, paint)
  }

  fun setTranslationX(translationX: Float) {
    this.translationX = translationX
  }

  fun setTranslationY(translationY: Float) {
    this.translationY = translationY
  }
}
